File: src/main/java/com/freya02/botcommands/api/core/DefaultEmbedSupplier.java
package com.freya02.botcommands.api.core;

import com.freya02.botcommands.api.commands.prefixed.BaseCommandEvent;
import com.freya02.botcommands.api.core.service.annotations.BService;
import com.freya02.botcommands.api.core.service.annotations.InterfacedService;
import com.freya02.botcommands.api.core.service.annotations.ServiceType;
import net.dv8tion.jda.api.EmbedBuilder;
import org.jetbrains.annotations.NotNull;

/**
 * Interface for embeds requested by {@link BaseCommandEvent#getDefaultEmbed()}, aiming to reduce boilerplate.
 * <br>This embed is also used in the default help command.
 *
 * <p>
 * <b>Usage:</b> Register your instance as a service with {@link BService}, and a {@link ServiceType} of {@link DefaultEmbedSupplier}.
 *
 * @see InterfacedService @InterfacedService
 */
@InterfacedService(acceptMultiple = false)
public interface DefaultEmbedSupplier {
	class Default implements DefaultEmbedSupplier {
		@NotNull
		@Override
		public EmbedBuilder get() {
			return new EmbedBuilder();
		}
	}

	@NotNull
	EmbedBuilder get();
}

File: BotCommands-commands/text/src/main/kotlin/io/github/freya022/botcommands/api/core/DefaultEmbedSupplier.kt
package io.github.freya022.botcommands.api.core

import io.github.freya022.botcommands.api.commands.text.BaseCommandEvent
import io.github.freya022.botcommands.api.core.service.annotations.BService
import io.github.freya022.botcommands.api.core.service.annotations.InterfacedService
import net.dv8tion.jda.api.EmbedBuilder

/**
 * Interface for embeds requested by [BaseCommandEvent.getDefaultEmbed], aiming to reduce boilerplate.
 * This embed is also used in the default help command.
 *
 * Returns an empty [EmbedBuilder] by default.
 *
 * **Usage**: Register your instance as a service with [BService].
 *
 * @see InterfacedService @InterfacedService
 */
@InterfacedService(acceptMultiple = false)
fun interface DefaultEmbedSupplier {

    fun get(): EmbedBuilder
}
